File: lab3/src/go/internal/cbtree.go
package internal

import (
	"strings"
)

type CBTNode struct {
	Data  string
	Left  *CBTNode
	Right *CBTNode
}

type CompleteBinaryTree struct {
	Root      *CBTNode
	nodeCount int
}

func NewCBT() *CompleteBinaryTree {
	return &CompleteBinaryTree{}
}

func (t *CompleteBinaryTree) Insert(value string) {
	newNode := &CBTNode{Data: value}
	if t.Root == nil {
		t.Root = newNode
		t.nodeCount = 1
		return
	}

	queue := []*CBTNode{t.Root}
	for len(queue) > 0 {
		temp := queue[0]
		queue = queue[1:]

		if temp.Left == nil {
			temp.Left = newNode
			break
		} else {
			queue = append(queue, temp.Left)
		}

		if temp.Right == nil {
			temp.Right = newNode
			break
		} else {
			queue = append(queue, temp.Right)
		}
	}
	t.nodeCount++
}

func (t *CompleteBinaryTree) Search(value string) bool {
	if t.Root == nil {
		return false
	}
	queue := []*CBTNode{t.Root}
	for len(queue) > 0 {
		temp := queue[0]
		queue = queue[1:]

		if temp.Data == value {
			return true
		}
		if temp.Left != nil {
			queue = append(queue, temp.Left)
		}
		if temp.Right != nil {
			queue = append(queue, temp.Right)
		}
	}
	return false
}

func (t *CompleteBinaryTree) ToSlice() []string {
	var res []string
	if t.Root == nil {
		return res
	}
	queue := []*CBTNode{t.Root}
	for len(queue) > 0 {
		temp := queue[0]
		queue = queue[1:]
		res = append(res, temp.Data)
		if temp.Left != nil {
			queue = append(queue, temp.Left)
		}
		if temp.Right != nil {
			queue = append(queue, temp.Right)
		}
	}
	return res
}

func (t *CompleteBinaryTree) Print() string {
	if t.Root == nil {
		return "Empty Tree"
	}
	var sb strings.Builder
	t.printHelper(t.Root, "", false, &sb)
	return sb.String()
}

func (t *CompleteBinaryTree) printHelper(node *CBTNode, prefix string, isLeft bool, sb *strings.Builder) {
	if node == nil {
		return
	}
	marker := "└── "
	if isLeft {
		marker = "├── "
	}
	sb.WriteString(prefix + marker + node.Data + "\n")

	newPrefix := prefix
	if isLeft {
		newPrefix += "│   "
	} else {
		newPrefix += "    "
	}

	if node.Left != nil || node.Right != nil {
		t.printHelper(node.Left, newPrefix, true, sb)
		t.printHelper(node.Right, newPrefix, false, sb)
	}
}

func (t *CompleteBinaryTree) GetSize() int {
	return t.nodeCount
}

File: lab3/src/cpp/src/cbtree.cpp
#include "cbtree.hpp"

#include <sstream>

void CompleteBinaryTree::insert(const std::string& value) {
    CBTNode* newNode = new CBTNode(value);
    if (!root) {
        root = newNode;
        nodeCount = 1;
        return;
    }

    std::queue<CBTNode*> q;
    q.push(root);

    while (!q.empty()) {
        CBTNode* temp = q.front();
        q.pop();

        if (!temp->left) {
            temp->left = newNode;
            break;
        } else {
            q.push(temp->left);
        }

        if (!temp->right) {
            temp->right = newNode;
            break;
        } else {
            q.push(temp->right);
        }
    }
    nodeCount++;
}

bool CompleteBinaryTree::search(const std::string& value) const {
    if (!root) return false;
    std::queue<CBTNode*> q;
    q.push(root);
    while (!q.empty()) {
        CBTNode* temp = q.front();
        q.pop();
        if (temp->data == value) return true;
        if (temp->left) q.push(temp->left);
        if (temp->right) q.push(temp->right);
    }
    return false;
}

void CompleteBinaryTree::clear() {
    clearRecursive(root);
    root = nullptr;
    nodeCount = 0;
}

void CompleteBinaryTree::clearRecursive(CBTNode* node) {
    if (node) {
        clearRecursive(node->left);
        clearRecursive(node->right);
        delete node;
    }
}

std::vector<std::string> CompleteBinaryTree::toVector() const {
    std::vector<std::string> res;
    if (!root) return res;
    std::queue<CBTNode*> q;
    q.push(root);
    while (!q.empty()) {
        CBTNode* temp = q.front();
        q.pop();
        res.push_back(temp->data);
        if (temp->left) q.push(temp->left);
        if (temp->right) q.push(temp->right);
    }
    return res;
}

std::string CompleteBinaryTree::print() const {
    if (!root) return "Empty Tree\n";
    std::string result = "";
    printHelper(root, "", false, result);
    return result;
}

void CompleteBinaryTree::printHelper(CBTNode* node, std::string prefix, bool isLeft,
                                     std::string& result) const {
    if (!node) return;
    result += prefix + (isLeft ? "├── " : "└── ") + node->data + "\n";
    if (node->left || node->right) {
        std::string newPrefix = prefix + (isLeft ? "│   " : "    ");
        printHelper(node->left, newPrefix, true, result);
        printHelper(node->right, newPrefix, false, result);
    }
}